docs/architecture/adr-058-auto-generated-cli.md

```diff
@@ -0,0 +1,83 @@
+moving away from CLI-style integration tests (which are really end-to-end tests) towards narrower integration tests
+which exercise `MsgClient` and `QueryClient` directly. This might involve replacing the current end-to-end CLI
+tests with unit tests as there still needs to be some way to test these CLI commands for full quality assurance.
+
+## Decision
+
+To make module development simpler, we provide infrastructure - in the new [`client/v2`](https://github.com/cosmos/cosmos-sdk/tree/main/client/v2)
+go module - for automatically generating CLI commands based on protobuf definitions to either replace or complement
+handwritten CLI commands. This will mean that when developing a module, it will be possibly to skip both writing and
+testing CLI commands as that can all be taken care of by the framework.
+
+The basic design for automatically generating CLI commands is to:
+* create one CLI command for each `rpc` method in a protobuf `Query` or `Msg` service
+* create a CLI flag for each field in the `rpc` request type
+* for `query` commands call gRPC and print the response as protobuf JSON or YAML
+* for `tx` commands, create a transaction and apply common transaction flags
+
+In order to make the auto-generated CLI as easy to use (or easier) than handwritten CLI, we need to do custom handling
+of specific protobuf field types so that the input format is easy for humans:
+* `Coin`, `Coins`, `DecCoin`, and `DecCoins` should be input using the existing format (i.e. `1000uatom`)
+* it should be possible to specify an address using either the bech32 address string or a named key in the keyring
+* `Timestamp` and `Duration` should accept strings like `2001-01-01T00:00:00Z` and `1h3m` respectively
+* pagination should be handled with flags like `--page-limit`, `--page-offset`, etc.
+* it should be possible to customize any other protobuf type either via its message name or a `cosmos_proto.scalar` annotation
+
+At a basic level it should be possible to generate a command for a single `rpc` method as well as all the commands for
+a whole protobuf `service` definition. It should be possible to mix and match auto-generated and handwritten commands.
+
+## Consequences
+
+### Backwards Compatibility
+
+Existing modules can mix and match auto-generated and handwritten CLI commands so it is up to them as to whether they
+make breaking changes by replacing handwritten commands with slightly different auto-generated ones.
+
+### Positive
+
+* module developers will not need to write CLI commands
+* module developers will not need to test CLI commands
+
+### Negative
+
+### Neutral
+
+## Further Discussions
+
+At some point in the future, we should allow customization of the short and long usage strings for commands as
+well as allowing shorthand aliases for flags (ex. `-a` for `--amount`). We may also consider a way to specify that
```

Comment:
I feel like improving short & long descriptions, and user specified examples, should be a high priority item for auto-generating CLI

Reply:
+1. Or else the generated CLI will feel half-baked. Ideally the "open discussion" below can be settled in this ADR?

I really feel against adding cli proto options in the .proto files, because of obfuscation with bloat. I have a small preference for those options in go code, but also okay with a separate config file.

Reply:
I agree. I think this should be high priority and we might want to hold off on using auto-generated CLI commands until there is support for short and long descriptions.

We should also include example here as well. I think this is an underutilized cobra command option and `cobra/doc` package separates example into its own section when auto-generating CLI documentation.

Reply:
I also have a preference for a separate config file or in go code.